### PR TITLE
DEV-1017 Support Trusting Tokens From Multiple Issuers

### PR DESCRIPTION
**Pull request by vincewhite**
# Description

This change allows our servers to trust JWT tokens that are issued/signed by different issuers. 

Changes to this repo include adding a new class `JWTMultiIssuerVerifierBuilder` that when provided a token returns a `com.auth0.jwt.JWTVerifier` based on the token's issuer ("iss" JWT field). 

This is now used by `com.zepben.auth.server.JWTAuthenticator` to perform token verification (which is used by EWB server (vertx)). 

The `JWTVerifier`'s returned by `JWTMultiIssuerVerifierBuilder` can also be easily used by the default? `JWTAuthenticationProvider` in our ktor app's e.g. EAS. 

`AuthRoute.routeFactory` has a new signature.

This shouldn't require any code changes/impact anything client side.

# Associated tasks

Required for update to EWB https://github.com/zepben/energy-workbench-server/pull/149

Required for update to EAS https://github.com/zepben/evolve-app-server/pull/96

# Test Steps

If you want to test by hand (without writing code) you can pull down the updated EWB and EAS branches:
* Check that they can both require a list of issues in their configs
* Get the config for our multi-organisation Azure Application `b0a9461f-4641-479a-aaac-764be0b42662` from myself or Zara.
* Make sure the evolve-wab-app config includes the following `domain` value: 
```  "auth": {
    "authType": "entraid",
    "domain": "https://login.microsoftonline.com/organizations/v2.0",
```
* Test that you can login and make requests from your own zeben account and the test user from Zara's test tenant.
* Remove one of the issues from the list of trusted issues in the EAS config. Confirm the user from that tenant/issuer(yourself, or the test user) can longer login/make requests to EAS, while the other user continues to work.
* Confirm that authentication with the Auth0 test account still works when Auth0 is configured (rather than entraid)
* These tests steps are bad and I feel bad, but I will help anyone set this up for testing this on their machine (since this functionality is also somewhat tedious to test via the SDK as well)

# Checklist

If any of these are not applicable, strikethrough the line `~like this~`. **Do not delete it!**. Let the reviewer decide if you should have done it.

### Code
- [x] I have performed a self review of my own code (including checking issues raised when creating the PR).
- [x] I have added/updated unit tests for these changes, and if not I have explained why they are not necessary.
- [x] I have commented my code in any hard-to-understand or hacky areas.
- [x] I have handled all new warnings generated by the compiler or IDE. **I'm ignoring expectCacheRefresh is always true**  
- [x] I have rebased onto the target branch (usually main).

### Documentation
- [x] I have updated the changelog.
- [ ] ~I have updated any documentation required for these changes.~ Functionality and configuration changes to support this in EWB and EAS will be documented in their repos. 

# Breaking Changes
- [x] I have considered if this is a breaking change and will communicate it with other team members if so.

Breaks basically anything that relies on this repo for server side authentication/authorization needs to be updated to support the new function signatures.

It's imported by `ewb-network-server` but I don't think it uses it for any auth?

`AuthRoute.routeFactory` has a new signature.

# Other things I touched
Reduced the leeway used when verifying timestamps in the JWT token (e.g. expiry date) from ~2 hours to 60 seconds.

`JWTAuthoriser.authorise` no longer accepts a permissions claims field, instead it will attempt to retrieve claims from the "permissions" field if it exists in the token, or the "roles" field if the "permissions" field doesn't exist.

### CODE REVIEW

**Comment by ryanjcoleman** [comment]
Task linked: [DEV-1017 Evolve-conn](https://app.clickup.com/t/6929263/DEV-1017)

File: pom.xml [hunk 1]
```diff
@@ -10,7 +10,7 @@
 
     <groupId>com.zepben.evolve</groupId>
     <artifactId>evolve-conn</artifactId>
-    <version>0.11.0-SNAPSHOT2</version>
+    <version>0.11.0-SNAPSHOT888</version>
```
Comment (zarakay) [inline pom.xml:13]: TODO remove?

**Review by zarakay — COMMENTED** [review]
minor things

File: .idea/codeStyles/Project.xml [hunk 2]
```diff
@@ -1,5 +1,6 @@
 <component name="ProjectCodeStyleConfiguration">
   <code_scheme name="Project" version="173">
+    <option name="RIGHT_MARGIN" value="160" />
```
[the remark this answers is not in the document]
Comment (kgreav) [reply .idea/codeStyles/Project.xml]: ftr we have only a few settings we set, and they should be reflected in here, but jetbrains isn't perfect. This is a correct setting, but like Vince said it was already in there but maybe different versions of IDE look up different values or something/maybe the IDE just gets confused occasionally. Most of the time it's safe to just leave these changes in if they look reasonable.

File: src/main/kotlin/com/zepben/auth/server/JWTAuthoriser.kt [hunk 3]
```diff
@@ -24,33 +24,35 @@ object JWTAuthoriser {
...
      *
      * @param token The JWT
      * @param requiredClaim The claim to authorise.
-     * @param permissionsField The field to extract claims from. Defaults to "permissions".
      */
     @JvmStatic
-    fun authorise(token: DecodedJWT, requiredClaim: String, permissionsField: String = "permissions"): AuthResponse {
-        val permissions = token.getClaim(permissionsField).asList(String::class.java).toHashSet()
+    fun authorise(token: DecodedJWT, requiredClaim: String): AuthResponse {
+        val permissions = run {
```
Comment (zarakay) [inline src/main/kotlin/com/zepben/auth/server/JWTAuthoriser.kt:34]: return authorize(token, setOf(requiredClaim))

**Pull request merged**